src/Data/Array/Accelerate/Type.hs

```diff
@@ -252,59 +250,131 @@ class IsScalar a => IsSingle a where
+  TypeWord   -> "Word"
+  TypeWord8  -> "Word8"
+  TypeWord16 -> "Word16"
+  TypeWord32 -> "Word32"
+  TypeWord64 -> "Word64"
+showSingleType (NumSingleType (FloatingNumType tp)) = case tp of
+  TypeHalf   -> "Half"
+  TypeFloat  -> "Float"
+  TypeDouble -> "Double"
+showSingleType (NonNumSingleType TypeChar) = "Char"
+showSingleType (NonNumSingleType TypeBool) = "Bool"
+
+-- Common used types in the compiler.
+scalarTypeBool :: ScalarType Bool
+scalarTypeBool = SingleScalarType $ NonNumSingleType TypeBool
+
+scalarTypeInt :: ScalarType Int
+scalarTypeInt = SingleScalarType $ NumSingleType $ IntegralNumType TypeInt
+
+scalarTypeInt32 :: ScalarType Int32
+scalarTypeInt32 = SingleScalarType $ NumSingleType $ IntegralNumType TypeInt32
+
+scalarTypeWord8 :: ScalarType Word8
+scalarTypeWord8 = SingleScalarType $ NumSingleType $ IntegralNumType TypeWord8
+
+scalarTypeWord32 :: ScalarType Word32
+scalarTypeWord32 = SingleScalarType $ NumSingleType $ IntegralNumType TypeWord32
+
+-- Tuple representation
 -- -------------------
 --
--- Representation of product types, consisting of:
+-- Both arrays (Acc) and expressions (Exp) may form tuples. These are represented
+-- using as product types, consisting of:
 --
 --   * unit (void)
 --
---   * scalar types: values which go in registers. These may be single value
+--   * single array / scalar types
+--     in case of expressions: values which go in registers. These may be single value
 --     types such as int and float, or SIMD vectors of single value types such
 --     as <4 * float>. We do not allow vectors-of-vectors.
 --
 --   * pairs: representing compound values (i.e. tuples) where each component
 --     will be stored in a separate array.
 --
-data TupleType a where
-  TypeRunit   ::                               TupleType ()
-  TypeRscalar :: ScalarType a               -> TupleType a
-  TypeRpair   :: TupleType a -> TupleType b -> TupleType (a, b)
+data TupR s a where
+  TupRunit   ::                         TupR s ()
+  TupRsingle :: s a                  -> TupR s a
+  TupRpair   :: TupR s a -> TupR s b -> TupR s (a, b)
+
+type TupleType = TupR ScalarType -- Rename to EltR?
+
+instance Show (TupR ScalarType a) where
+  show TupRunit       = "()"
+  show (TupRsingle t) = show t
+  show (TupRpair a b) = "(" ++ show a ++ "," ++ show b ++")"
+
+type Tup2 a b               =        (((), a), b)
+type Tup3 a b c             =       ((((), a), b), c)
+type Tup4 a b c d           =      (((((), a), b), c), d)
+type Tup5 a b c d e         =     ((((((), a), b), c), d), e)
+type Tup6 a b c d e f       =    (((((((), a), b), c), d), e), f)
+type Tup7 a b c d e f g     =   ((((((((), a), b), c), d), e), f), g)
+type Tup8 a b c d e f g h   =  (((((((((), a), b), c), d), e), f), g), h)
+type Tup9 a b c d e f g h i = ((((((((((), a), b), c), d), e), f), g), h), i)
+type Tup16 a b c d e f g h
+           i j k l m n o p   = ((((((((((((((((((), a), b), c), d), e), f), g), h), i), j), k), l), m), n), o), p))
```

Comment:
I believe there is a redundant pair of outer parentheses here :)

Reply:
Thanks, good catch!